docs/design/enums_or_ids.md

```diff
@@ -0,0 +1,129 @@
+
+1. Is the set **fixed**?
+2. Is the set **comprehensive**? Is it inclusive of all peoples and cultures? Are there no additional entities that are not currently represented?
+3. Are there **30 or fewer items** in the set?
+4. Are the entities used within the context of a single program, and **not normally interchanged** between different programs?
+5. Are entities in the set usually **compared** with one another, and rarely expressed in a standalone context?
+
+If you answered *yes* to the above questions, then an enum is probably the right choice. If you answered *no* to the above questions, then an identifier is probably the right choice. If there is a mix of *yes* and *no*, you are in a gray area; continue reading this document for additional guidance.
+
+## Enumerations
+
+An *enumeration* is a fixed list of entities. In Rust, an enumeration looks like this:
+
+```rust
+enum FooEnum {
+    Entity1,
+    Entity2,
+    Entity3,
+}
+```
+
+### Advantages
+
+- Enums are useful for large codebase maintenance. When an enum changes, engineers can identify and update all call sites.
+- The compiler has many opportunities to optimize enums; for example, memory layout and match statements.
+
+### Examples
+
+Examples of entities that are represented by enums:
+
+1. [`Signum`](https://unicode-org.github.io/icu4x-docs/doc/fixed_decimal/signum/enum.Signum.html):
+    - Set of 4 values, determined by ECMA-402
+    - Unlikely to change over time
+    - Does not directly relate to peoples and cultures
+2. [`PluralCategory`](https://unicode-org.github.io/icu4x-docs/doc/icu_plurals/enum.PluralCategory.html):
+    - Set of 6 values, determined by Unicode/CLDR
+    - Could theoretically change, but has not grown for a long period of time
+    - Correctly represents all languages supported by CLDR
+
+### Private-use variants
+
+An enum can have a private-use variant:
+
+```rust
+enum FooEnum {
+    Entity1,
+    Entity2,
+    Entity3,
+    PrivateUse(TinyStr8),
+}
+```
+
+An enum with a private-use variant bridges the extensibility of identifiers with the engineering convenience of enums.
+
+### Explicit discriminants
+
+Enum values can be given explicit discriminant values:
+
+```rust
+enum FooEnum {
+    Entity1 = 1,
+    Entity2 = 2,
+    Entity3 = 3,
+}
+```
+
+If an enum is used for interchange between programs, including over FFI, discriminants should be explicitly specified. However, if the enum is only used within the context of the current program (if it never leaves Rust), explicitly specifying the discriminants could limit the compiler's ability to optimize the enum.
+
+### Non-exhaustive enums
+
+It is possible in Rust to mark enums as `#[non_exhaustive]`, which allows additional entries to be added over time.
+
+However, if there is a need to do this, then it means that the set of entities could grow over time. If this is the case, you should probably be using identifiers instead.
```

Comment:
this sentence seems to indicate that option bag values (which you wanted to have as non exhaustive) should be identifiers, which I disagree with.

Reply:
Option bags are not sets of entities.  This doc is only about sets of entities.

Reply:
I weakened the statement.

Reply:
Option bag options values may be:

```rust
#[non_exhaustive]
enum DateLength {
  Short,
  Medium,
  Long
}

struct Options {
  date: DateLength
}
```

Reply:
> I weakened the statement.

The statement is still stronger than my opinion. It seems to nudge reader in the direction of not using enums for situations like in the example above where I claim enum to be a better choice than identifier.

I'm ok with merging this PR as is, I don't think we'll always agree on such topics so moving toward my position would likely start pushing this away from yours.

Reply:
Let's look at DateLength and put it on the 5 point decision tree.

1. Is the set **fixed**?
    - MAYBE; it grows fairly rarely
2. Is the set **comprehensive**? Is it inclusive of all peoples and cultures? Are there no additional entities that are not currently represented?
    - YES; it is defined by CLDR in order to cover all peoples and cultures
3. Are there **30 or fewer items** in the set?
    - YES
4. Are the entities used within the context of a single program, and **not normally interchanged** between different programs?
    - YES
5. Are entities in the set usually **compared** with one another, and rarely expressed in a standalone context?
    - YES

So you can see that DateLength clearly favors Enum, even though we applied `#[non_exhaustive]` out of an abundance of caution.

Do you have a concrete suggestion on how to improve this language further, or are you satisfied?

Reply:
okay I also weakened the paragraph again.  PTAL and approve if you like the doc.